### PR TITLE
Add a few more merge-related bindings

### PR DESCRIPTION
**Pull request by tksfz**
Adding repo.merge_commits() here, repo.cleanup_state(), as well as ref.set_target()

### CODE REVIEW

File: src/reference.rs [hunk 1]
```diff
@@ -179,6 +179,19 @@ impl<'repo> Reference<'repo> {
         }
     }
 
+    /// Conditionally create a new reference with the same name as the given
+    /// reference but a different OID target. The reference must be a direct
+    /// reference, otherwise this will fail.
```
Comment (alexcrichton) [inline src/reference.rs]: Can this also include the clause "The new reference will be written to disk, overwriting the given reference."

File: src/reference.rs [hunk 2]
```diff
@@ -179,6 +179,19 @@ impl<'repo> Reference<'repo> {
...
+    pub fn set_target(&mut self, id: Oid, msg: &str) -> Result<Reference<'repo>, Error> {
```
Comment (alexcrichton) [inline src/reference.rs]: Could you try passing in a `None` here for `msg` and see what happens? It wasn't clear to me whether it was required or not from reading the code.

Comment (alexcrichton) [inline src/reference.rs]: er, reading the libgit2 code that is

File: src/repo.rs [hunk 3]
```diff
@@ -1041,6 +1041,32 @@ impl Repository {
         Ok(())
     }
 
+    /// Merge two commits, producing a git_index that reflects the result of
+    /// the merge. The index may be written as-is to the working directory or
+    /// checked out. If the index is to be converted to a tree, the caller
+    /// should resolve any conflicts that arose as part of the merge.
```
Comment (alexcrichton) [inline src/repo.rs]: Can this also include the clauses that start with "The merge performed uses the first common ancestor ..."

File: src/repo.rs [hunk 4]
```diff
@@ -1041,6 +1041,32 @@ impl Repository {
         Ok(())
     }
 
+    /// Merge two commits, producing a git_index that reflects the result of
```
Comment (alexcrichton) [inline src/repo.rs]: s/a git_index/an Index/

**Comment by alexcrichton** [comment]
Overall looks great to me, thanks @tksfz!

File: src/repo.rs (hunk 4 above)
Comment (tksfz) [inline src/repo.rs]: done

File: src/repo.rs (hunk 3 above)
Comment (tksfz) [inline src/repo.rs]: done

File: src/reference.rs (hunk 1 above)
Comment (tksfz) [inline src/reference.rs]: done

File: src/reference.rs (hunk 2 above)
Comment (tksfz) [inline src/reference.rs]: So when I pass in `0 as *const libc::c_char`  for msg (to the raw call), everything seems to work (in at least one particular case).  The reflog records an empty message.

Do you  want  me to make `msg` an `Option`?  I could do that but then we might want  to  modify the  existing `reference.rename()` function, which also takes a  msg argument, to be the same.

Comment (tksfz) [inline src/reference.rs]: IMHO I'd be inclined to leave it as `msg: &str`.  If people want an empty log message they can/should pass in empty string.

**Comment by tksfz** [comment]
Btw  check the conversation here for reference.rs (showing up as hidden for an outdated diff):

So when I pass in `0 as *const libc::c_char`  for msg (to the raw call), everything seems to work (in at least one particular case).  The reflog records an empty message.

Do you  want  me to make `msg` an `Option`?  I could do that but then we might want  to  modify the  existing `reference.rename()` function, which also takes a  msg argument, to be the same.

IMHO I'd be inclined to leave it as `msg: &str`.  If people want an empty log message they can/should pass in empty string.

**Comment by alexcrichton** [comment]
Thanks!

**Comment by alexcrichton** [comment]
(not having an `Option` is fine)

**Pull request merged**